pyop2/host.py

```diff
@@ -248,60 +250,74 @@ def c_addto_scalar_field(self, i, j, buf_name, extruded=None, is_facet=False):
-                       'r': r,
-                       'elem_idx': "i * %d +" % (nrows) if xtr == "" else "",
-                       'xtr': xtr}
-                col = "%(m)s * %(xtr)s%(map)s[%(elem_idx)si_1] + %(c)s" % \
-                      {'m': cmult,
-                       'map': self.c_map_name(1, j),
-                       'dim': ncols,
-                       'c': c,
-                       'elem_idx': "i * %d +" % (ncols) if xtr == "" else "",
-                       'xtr': xtr}
-
-                s.append('addto_scalar(%s, %s, %s, %s, %d)'
-                         % (self.c_arg_name(i, j), val, row, col, self.access == WRITE))
-        return ';\n'.join(s)
+        if is_facet:
+            nrows *= 2
+            ncols *= 2
+
+        ret = []
+        rbs, cbs = self.data.sparsity[i, j].dims[0][0]
+        rdim = rbs * nrows
+        cdim = cbs * ncols
+        buf_name = buf[0]
+        tmp_name = layout[0]
+        addto_name = buf_name
+        addto = 'MatSetValuesLocal'
+        if uses_layout:
+            # Padding applied, need to pack into "correct" sized buffer for matsetvalues
+            ret = ["""for ( int j = 0; j < %(nrows)d; j++) {
+                          for ( int k = 0; k < %(ncols)d; k++ ) {
+                              %(tmp_name)s[j][k] = %(buf_name)s[j][k];
+                          }
+                      }""" % {'nrows': rdim,
+                              'ncols': cdim,
+                              'tmp_name': tmp_name,
+                              'buf_name': buf_name}]
+            addto_name = tmp_name
+        if self.data._is_vector_field:
+            addto = 'MatSetValuesBlockedLocal'
+            if self._flatten:
+                if applied_blas:
+                    idx = "[(%%(ridx)s)*%d + (%%(cidx)s)]" % rdim
+                else:
+                    idx = "[%(ridx)s][%(cidx)s]"
+                ret = []
+                idx_l = idx % {'ridx': "%d*j + k" % rbs,
+                               'cidx': "%d*l + m" % cbs}
+                idx_r = idx % {'ridx': "j + %d*k" % nrows,
+                               'cidx': "l + %d*m" % ncols}
+                # Shuffle xxx yyy zzz into xyz xyz xyz
+                ret = ["""
+                for ( int j = 0; j < %(nrows)d; j++ ) {
+                   for ( int k = 0; k < %(rbs)d; k++ ) {
+                      for ( int l = 0; l < %(ncols)d; l++ ) {
+                         for ( int m = 0; m < %(cbs)d; m++ ) {
+                            %(tmp_name)s%(idx_l)s = %(buf_name)s%(idx_r)s;
+                         }
+                      }
+                   }
+                }""" % {'nrows': nrows,
+                        'ncols': ncols,
+                        'rbs': rbs,
+                        'cbs': cbs,
+                        'idx_l': idx_l,
+                        'idx_r': idx_r,
+                        'buf_name': buf_name,
+                        'tmp_name': tmp_name}]
+                addto_name = tmp_name
```

Comment:
This code generation strategy is not  a thing of beauty. But that's a more general host.py problem to be fixed some other time.